routing_plugin.go

```diff
@@ -104,3 +120,419 @@ func ReleaseRoutingPlugin(ctx context.Context) {
+				visited[peer] = struct{}{}
+				peers = append(peers, peer)
+			}
+		}
+
+		if len(peers) == 1 {
+			exit = target
+			target = peers[0]
+			continue
+		}
+
+		// If there are no more peers to visit then we can't use
+		// this routing plugin.
+		if len(peers) == 0 {
+			return ErrRoutingPluginNotApplicable
+		}
+
+		// Found the first fork to our target.
+		break
+	}
+
+	nodeInfo, err := getNodeInfo(ctx, r.lnd.Client, target, privateEdges)
+	if err != nil {
+		return err
+	}
+
+	log.Debugf("Routing plugin target: '%v' %v", nodeInfo.Alias,
+		nodeInfo.PubKey.String())
+
+	// Gather node info (including channels) for the nodes we're
+	// interested in.
+	targetChanged := exit != target
+	nodes := make(map[route.Vertex]*lndclient.NodeInfo)
+	for _, edge := range nodeInfo.Channels {
+		// Skip edges to the exit node.
+		if targetChanged && (edge.Node1 == exit ||
+			edge.Node2 == exit) {
+			continue
+		}
+
+		var node route.Vertex
+		if edge.Node1 == target {
+			node = edge.Node2
+		} else {
+			node = edge.Node1
+		}
+
+		nodeInfo, err := getNodeInfo(
+			ctx, r.lnd.Client, node, privateEdges,
+		)
+		if err != nil {
+			return err
+		}
+
+		nodes[node] = nodeInfo
+	}
+
+	// Save MC state.
+	entries, err := r.lnd.Router.QueryMissionControl(ctx)
+	if err != nil {
+		return err
+	}
+	r.mcState = entries
+	log.Debugf("Saved MC state: %v", spew.Sdump(r.mcState))
+
+	// maxFeePerNode assigns the maximum fees that would be paid through
+	// selected nodes to the target node for the target amount.
+	maxFeePerNode := make(map[route.Vertex]int64)
+	totalCapacityPerNode := make(map[route.Vertex]btcutil.Amount)
+
+	amtMsat := int64(amt * 1000)
+	for nodeID, node := range nodes {
+		var totalCapacity btcutil.Amount
+		for _, ch := range node.Channels {
+			var policy *lndclient.RoutingPolicy
+			if ch.Node1 == nodeID && ch.Node2 == target {
+				policy = ch.Node1Policy
+			} else if ch.Node1 == target && ch.Node2 == nodeID {
+				policy = ch.Node2Policy
+			}
+
+			if policy == nil {
+				continue
+			}
+
+			totalCapacity += ch.Capacity
+
+			log.Debugf("'%v', policy=%v",
+				node.Alias, spew.Sdump(policy))
+			fee := policy.FeeBaseMsat +
+				policy.FeeRateMilliMsat*amtMsat
+
+			// For all peers we'll save the "maximum" routing fee
+			// for that peer (given the 1mm sats dummy amount).
+			if fee > maxFeePerNode[nodeID] {
+				maxFeePerNode[nodeID] = fee
+			}
+		}
+		totalCapacityPerNode[nodeID] = totalCapacity
+	}
+
+	r.nodesByMaxFee = make([]nodeFeeInfo, 0, len(maxFeePerNode))
+
+	// Sort peers by maximum fee, so that we can later or disable edges
+	// in a gradual way.
+	for nodeID, fee := range maxFeePerNode {
+		r.nodesByMaxFee = append(
+			r.nodesByMaxFee, nodeFeeInfo{
+				node:     nodeID,
+				capacity: totalCapacityPerNode[nodeID],
+				fee:      fee,
+			},
+		)
+	}
+
+	sort.Slice(r.nodesByMaxFee, func(i, j int) bool {
+		return r.nodesByMaxFee[i].fee < r.nodesByMaxFee[j].fee
+	})
+
+	if log.Level() == btclog.LevelTrace {
```

Comment:
Why this check? To avoid going through the slice below?

Reply:
Removed, yeah just to not iterate the slice but it's unimportant really.